api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirServiceRequestDaoImpl.java

```diff
@@ -83,5 +101,54 @@ private void handleCodedConcept(Criteria criteria, TokenAndListParam code) {
 		    Optional.of(or(toCriteriaArray(Stream.of(handleDate("dateStopped", dateRangeParam.getUpperBound()),
 		        handleDate("autoExpireDate", dateRangeParam.getUpperBound())))))))));
 	}
+
+	private void handleHasAndParam(Criteria criteria, HasAndListParam hasAndListParam) {
+		if (hasAndListParam != null) {
+		
+		if (lacksAlias(criteria, "obss")) {
+				criteria.createAlias("obs", "aliasObss");
+			}
+
+		DetachedCriteria detachedCriteria = DetachedCriteria.forClass(Obs.class);
+
+		 handleAndListParam(hasAndListParam, param -> {
+			if ( param.getValueAsQueryToken(fhirContext) != null) {
+				if (lacksAlias(criteria, "o")) {
+					criteria.createAlias("aliasObss.category", "o");
```

Comment:
So this is going to try to generate something like:

```sql
INNER JOIN aliasObss_.category o
```

Which probably isn't going to work. Is there a way we could maybe call `setupSearchParams()` in the `FhirObsDao`?

Reply:
@ibacher are you talking about [setupSearchParam(](https://github.com/openmrs/openmrs-module-fhir2/blob/3a085ce12511f1efae4f99d6b6ed54f950975099/api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirObservationDaoImpl.java#L44)) in `FhirObservationImpl `

Secondly i think the `INNER JOIN aliasObss_.category o` can be avoided by doing somthing like 
```
if (lacksAlias(criteria, "obs")) {
	criteria.createAlias("obs", "obs1");
			
}
```

Reply:
So here's what I was thinking: the way in which we handle the FHIR `Observation.category` is a bit... unique (the implementation is [here](https://github.com/openmrs/openmrs-module-fhir2/blob/495c6666e51c1d2f7416a7b0c6f729fe8d8d98b6/api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirObservationDaoImpl.java#L134); note that it involves creating a subquery and then checking if the `Obs.conceptClass.uuid` field is in a set returned from `categoryMap.queryConceptClassByCategory()`. That's a lot of logic to replicate here. Plus, we might want to filter the `Obs` based on other properties rather than just `category`. So the question is is there some way we can reuse the logic that already exists in `FhirObservationDaoImpl` here?

Reply:
What we ultimately want to end up with is something like this query:

```sql
SELECT *
FROM drug_order do_
JOIN obs ob_ ON ob_.order_id = do_.order_id
JOIN concept c_ ON ob_.concept_id = c_.concept_id
JOIN concept_class cc_ on  c_.class_id = cc_.concept_class_id
WHERE cc_.uuid IN (?, ?, ?)
```

Or, somewhat equivalently:

```sql
SELECT *
FROM drug_order do_
WHERE do_.id IN (
    SELECT ob_.order_id
    FROM obs ob_
    JOIN concept c_ ON ob_.concept_id = c_.concept_id
    JOIN concept_class cc_ on  c_.class_id = cc_.concept_class_id
	WHERE cc_.uuid IN (?, ?, ?)
)
```

Reply:
I'm not exactly sure how to generate the first of those (it's a little tricky because the linkage is only one way), but the second could be created by something like:

```java
Criteria criteria = sessionFactory.getCurrentSession().createCriteria(DrugOrder.class);
DetachedCriteria criteria2 = DetachedCriteria.forClass(Obs.class, "obs");
criteria2.createAlias("concept", "c");
criteria2.createAlias("c.conceptClass", "cc");
criteria2.add(Subqueries.propertyIn("cc.uuid", categoryMap.queryConceptClassByCategory("laboratory")));
criteria.add(Subqueries.propertyIn("id", criteria2));
```

So the question is whether we can (somehow) use `FhirObservationDaoImpl#setupSearchParams()` to populate the criteria for `criteria2` above and somewhat simplify the code we need to implement in the `FhirServiceRequestDaoImpl` class.

Reply:
hello @ibacher
am sorry for the delayed response ,,i have been looking into the commnets you made above and your suggestion, i would say implementing`  FhirObservationDaoImpl#setupSearchParams(`) in `FhirServiceRequestDaoImpl `would work since it will cover the Observation.category issue with the concept class mapping plus also catering the other filtering on the obs ie patient, encounter references something like `/ServiceRequest?_has:Observation:based-on:patient=uuid `(please correct me if am wrong on this). The tricky bit here has been implementing `FhirObservationDaoImpl#setupSearchParams()` after `Auto-wiring` it in the `FhirServiceRequestDaoImpl `class because of its void nature and taking in a `SearchParameterMap` argument